### DISCUSSION TITLE
Transitioning the virtualisation library to a pre-existing solution rather than using Sedna

### DISCUSSION THREAD

**Discussion by DecDuck** [fnuecke/oc2#281] [General]
From my understanding, development has stopped due to some issues with the underlying RISC-V library, Sedna. I wanted to chat about potentially moving to different library/solution to restart development or move it to active developers.

**Comment by DecDuck** [comment]
As mentioned in the Sedna issue, the issue does not persist in QEMU (See https://github.com/fnuecke/sedna/issues/12). As a somewhat unlikable alternative to pure-Java emulation, it's possible to use a QEMU client (something like https://github.com/shevek/qemu-java) instead of running the emulator within Java.

**Reply by untodesu** [reply]
The main issue with using a separate thing would be the need for it to be installed alongside the game, in case of a mod with a RV emulator packed in you can just install it and run it on any OS without worrying about anything else.

I'd suggest not using QEMU but instead building a RISC-V VM in, say C++ and linking it with Java through JNI. This would still require OS-specific binaries but is simpler than using QEMU

**Reply by DecDuck** [reply]
I suggested QEMU because it's a known-good RISC-V VM. Honestly, a generic interface could be created in Java and then have a variety of implementations depending on what's installed.

**Reply by DecDuck** [reply]
That might actually be the first step - migrating to a generic interface and then moving Sedna to that interface.

**Reply by untodesu** [reply]
Good idea, I guess

**Reply by donmor** [reply]
Maybe we can just let the mod download prebuilt qemu binaries during game initialization, if executable `qemu-system-riscv64` is not detected in PATH.

**Reply by donmor** [reply]
BTW I wonder if we can have raw images stored in NBT and provided to qemu as NBD via sockets.
